pennylane/circuit_graph.py

```diff
@@ -439,84 +403,74 @@ def update_node(self, old, new):
-        self._graph[self._indices[id(old)]] = new
-        index = self._indices.pop(id(old))
-        self._indices[id(new)] = index
+        self._inds_for_objs[WrappedObj(new)] = self._inds_for_objs.pop(WrappedObj(old))
 
-        self._operations = self.operations_in_order
-        self._observables = self.observables_in_order
+        for i, op in enumerate(self._operations):
+            if op is old:
+                self._operations[i] = new
+        for i, mp in enumerate(self._observables):
+            if mp is old:
+                self._observables[i] = new
+        for i, obj in enumerate(self._queue):
+            if obj is old:
+                self._queue[i] = new
 
     def get_depth(self):
         """Depth of the quantum circuit (longest path in the DAG)."""
+        return self._depth
+
+    @cached_property
+    def _depth(self):
         # If there are no operations in the circuit, the depth is 0
         if not self.operations:
-            self._depth = 0
-
-        # If there are operations but depth is uncomputed, compute the truncated graph
-        # with only the operations, and return the longest path + 1 (since the path is
-        # expressed in terms of edges, and we want it in terms of nodes).
-        if self._depth is None and self.operations:
-            if self._operation_graph is None:
-                self._operation_graph = self._graph.subgraph(
-                    list(self._indices[id(node)] for node in self.operations)
-                )
-                self._extend_graph(self._operation_graph)
-                self._depth = (
-                    rx.dag_longest_path_length(
-                        self._operation_graph, weight_fn=lambda _, __, w: self._weight_func(w)
-                    )
-                    + 1
-                )
-        return self._depth
+            return 0
+        with QueuingManager.stop_recording():
+            ops_with_initial_I = [
+                I(self.wires)
+            ] + self.operations  # add identity wire to end the graph
+        operation_graph, _, _ = _construct_graph_from_queue(ops_with_initial_I, self.wires)
+
+        # pylint: disable=unused-argument
+        def weight_fn(in_idx, out_idx, w):
+            out_op = ops_with_initial_I[out_idx]
+            if isinstance(out_op, ResourcesOperation):
+                return out_op.resources().depth
+            return 1
+
+        return rx.dag_longest_path_length(operation_graph, weight_fn=weight_fn)
+
+    def has_path_idx(self, a_idx: int, b_idx: int) -> bool:
+        """Checks if a path exists between the two given nodes.
 
-    @staticmethod
-    def _weight_func(weight):
-        """If weight is a number, use it!"""
-        if isinstance(weight, Number):
-            return weight
-        return 1
+        Args:
+            a (int): initial node index
+            b (int): final node index
 
-    def _extend_graph(self, graph: rx.PyDiGraph) -> rx.PyDiGraph:
```

Comment:
previously used in calculating the weight for the depth graph. No longer needed.